onecodex/input_helpers.py

```diff
@@ -3,51 +3,140 @@
+PAIRED_ORDINAL_REV_PATTERN = r"([._][Rr])\d+([._][\w.]+)$"
+
+log = logging.getLogger("onecodex")
+
+
+def _replace_filename_ordinal(filename, replacement):
+    """Replace file[_.]num[._] with file[_.]replacement[._]."""
+    replace_pattern = rf"\g<1>{replacement}\g<2>"
+    return re.sub(ORDINAL_REV_PATTERN, replace_pattern, filename)
+
+
+def _replace_paired_filename_ordinal(filename, replacement):
+    """Replace file[_.][Rr]?num[._] with file[_.][Rr]?replacement[._]."""
+    first_pass = _replace_filename_ordinal(filename, replacement)
+    replace_pattern = rf"\g<1>{replacement}\g<2>"
+    return re.sub(PAIRED_ORDINAL_REV_PATTERN, replace_pattern, first_pass)
+
+
+def concatenate_ont_groups(files, prompt):
+    """Concatenate ONT split files and return the group as a single entry on the files list."""
+    single_files = set(files)
+    ont_groups = defaultdict(set)
+    auto_group = True
+
+    for filename in files:
+        ont_zero_filename = _replace_filename_ordinal(filename, "0")
+        if os.path.exists(ont_zero_filename):
+            # strip the ordinal and preceding . or _
+            base_filename = re.sub(r"[._]\d([._][\D._]+)$", r"\1", filename)
+            base_filename = os.path.join(tempfile.gettempdir(), base_filename)
+
+            ont_groups[base_filename].add(filename)
+
+    # filter to groups of at least 2 files
+    ont_groups = {k: v for k, v in ont_groups.items() if len(v) >= 2}
+
+    if prompt and ont_groups:
+        group_list = ""
+        n_grouped_files = 0
+        for base_ont_filename, group in ont_groups.items():
+            n_grouped_files += len(group)
+            group_list += f"\n  {os.path.basename(base_ont_filename)}"
+        answer = click.confirm(
+            "It appears there are {n_grouped_files} ONT files (of {n_files} total):{group_list}\nConcatenate them before upload?".format(
+                n_grouped_files=n_grouped_files,
+                n_files=len(single_files),
+                group_list=group_list,
+            ),
+            default="Y",
+        )
+
+        if not answer:
+            auto_group = False
+
+    if not auto_group:
+        return files
+
+    # Ensure there is no gap in the file sequences
+    for base_ont_filename, files in ont_groups.items():
+        ont_file = next(iter(files))
+        expected_sequence = [_replace_filename_ordinal(ont_file, idx) for idx in range(len(files))]
+        if any(expected_file not in files for expected_file in expected_sequence):
+            # Detected a gap in the ONT file sequence, skipping
+            log.warning(
+                f"Detected a gap in the ONT file sequence for {os.path.basename(base_ont_filename)}, skipping concatenation"
+            )
+            continue
+        with open(base_ont_filename, "wb") as outf:
+            for ont_filename in expected_sequence:
+                with open(ont_filename, "rb") as inf:
+                    shutil.copyfileobj(inf, outf)
```

Comment:
I think this is fine we we are assuming gzipped. If they're not gzipped; what happens if there is no newline at the end of a file?

E.g.

```
@seq1
GATC
+seq1
xxxx@seq2
GATC
+seq2
xxxx
```

Reply:
I think we can skip that check as I see that by default, the ONT files do have newlines at the end and they concatenate just fine